src/render/render_quantum.rs

```diff
@@ -464,15 +470,10 @@ impl AudioRenderQuantum {
         self.channels.iter_mut().for_each(fun)
     }
 
-    /// Sum two `AudioRenderQuantum`s
-    ///
-    /// If the channel counts differ, the buffer with lower count will be upmixed.
-    pub fn add(&mut self, other: &Self, interpretation: ChannelInterpretation) {
-        // mix buffers to the max channel count
-        let channels_self = self.number_of_channels();
-        let channels_other = other.number_of_channels();
-        let channels = channels_self.max(channels_other);
-
+    /// Sum two `AudioRenderQuantum`s, both quantums will be up/down mix according
+    /// to given number of channels and channel interpretation.
+    /// see <https://webaudio.github.io/web-audio-api/#channel-up-mixing-and-down-mixing>
+    pub fn add(&mut self, other: &Self, channels: usize, interpretation: ChannelInterpretation) {
```

Comment:
I think I would prefer not to take a `channel` argument here, but just assume `self` already has the right number of channels. To me that makes it more clean and also more semantical: if you want to add another buffer to an existing one, the number of channels will not change in the result

Looking at the code in `graph.rs`, the inputs are not already set to the right number of channels, so that should then be changed
```
    pub fn add_node(
        &mut self,
        index: NodeIndex,
        processor: Box<dyn AudioProcessor>,
        inputs: usize,
        outputs: usize,
        channel_config: ChannelConfig,
    ) {
        // todo, allocate on control thread, make single alloc..?
        let inputs = vec![AudioRenderQuantum::new(self.alloc.silence()); inputs];
        let outputs = vec![AudioRenderQuantum::new(self.alloc.silence()); outputs];
```

Reply:
Hum I think I see what you mean, but there was something strange I don't recall precisely, my impression was that `add` should be:
- either `add(&self, other)`, with all mixing done by the caller and `add` being completely dumb
- either `add(&self, other, num_channels, interpretation)`, with arbitrary input and `add` should take care of all mixing everything (which appeared more safe in the end than the dumb version)

But I don't really remember why I ended up thinking that, maybe that's a misunderstanding on my side (thinking too much of `add` as a kind of static method) I will review that and tell you if something comes back in the way

Reply:
Cool, let me know. My opinion is that `add` should be allowed to perform some mixing to align the buffers

Reply:
Hey, no I didn't have any time to continue last week, and this week will be probably be a bit busy too so if you have some time, yes that's a good idea to go on with the fix (feel free to push on this branch if you want too)!